### ISSUE TITLE
TOML: Missing dependency on Antlr4 runtime

### ISSUE THREAD

**Issue by dnault** [ConsenSys/cava#30]
Hi! Really nice TOML parsing library -- thanks for sharing it under a business-friendly license. I especially like the detailed messages for syntax validation errors. I thought I should report an issue I ran into while getting started.

In a project that uses `net.consensys.cava:cava-toml:0.2.0`, parsing a TOML config file fails with the following exception unless the Antlr4 runtime is added as an explicit dependency:

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/antlr/v4/runtime/CharStream
	at com.couchbase.connector.elasticsearch.Config.main(Config.java:147)
Caused by: java.lang.ClassNotFoundException: org.antlr.v4.runtime.CharStream
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 1 more
```

The workaround is to manually add a dependency on the Antlr4 runtime. For gradle, this looks like

```
compile 'net.consensys.cava:cava-toml:0.2.0'
compile 'org.antlr:antlr4-runtime:4.7.1'
```

It would be nice if the Maven POM for the `cava-toml` library expressed a transitive dependency on the Antlr4 runtime. Better yet, embed the runtime in the `cava-toml` library as a shaded/shadowed dependency.

**Comment by cleishm** [comment]
Hi @dnault,

This is something we need to think about for the cava project in general. Currently it takes a cautious line about requiring dependencies for the `net.consensys.cava:cava` library (which includes all the components), in order to avoid it pulling in dozens of libraries for projects that don't need them. That caution is reflected in the dependency definitions of individuals components and, in most cases, documented explicitly - e.g. https://github.com/ConsenSys/cava/blob/master/crypto/src/main/java/net/consensys/cava/crypto/sodium/package-info.java#L12-L14

This documentation is missing for the TOML component and should be added.

However, I acknowledge it would be preferable to have the dependency automatically added when including the component individually. I wonder if anyone has good ideas on how to do this with gradle/maven?

**Comment by cleishm** [comment]
I suppose another approach could be to remove the uber-library entirely and require projects taking advantage of cava to include each desired component explicitly. That would allow transitive dependencies to be included at the trade-off of making cava harder to consume. WDYT, @atoulme?

**Comment by atoulme** [comment]
Two interesting requirements:
Add `antlr-runtime` dependency to cava-toml
Make sure `antlr-runtime` doesn't get pulled into the uber-jar dependencies.

I think we could get away by listing the dependencies on the uber-jar, but mark them optional.
See https://maven.apache.org/guides/introduction/introduction-to-optional-and-excludes-dependencies.html

**Comment by cleishm** [comment]
Right - so we already mark them as optional in each component. E.g. `jrn-ffi` is marked as `compileOnly` for the cava-crypto component: https://github.com/ConsenSys/cava/blob/master/crypto/build.gradle#L10

But this is probably a good approach. Let's change those to non-optional, and then substitute them for optional dependencies when we create the uber-jar: https://github.com/ConsenSys/cava/blob/master/build.gradle#L343-L354

**Comment by cleishm** [comment]
Hi @dnault!

I believe this is resolved by #33. If you can verify using the snapshot distribution, that would be fantastic. The first snapshot including this change is `net.consensys.cava:cava:0.3.0-CD4FEE-snapshot`.

Cheers,
Chris

**Comment by dnault** [comment]
@cleishm Verified, thanks!

**Comment by dnault** [comment]
@cleishm I see you the new snapshot also removed the unnecessary dependency on `org.antlr:antlr4` and now just depends on the runtime. Nice!

**Comment by cleishm** [comment]
@dnault FYI, this was released in cava `0.3.0`
